File: rabbitmq/golang/utils/rabbitmq_util.go
package utils

import (
	"log"
	"time"

	"github.com/rabbitmq/amqp091-go"
)

type RabbitmqUtil interface {
	GetConnection() *amqp091.Connection
	GetChannel() *amqp091.Channel
	Close(host string, port string)
}

type rabbitmqUtil struct {
	Connection *amqp091.Connection
	Channel    *amqp091.Channel
}

func NewRabbitmqConnection(host string, username string, password string, port string) RabbitmqUtil {
	println(time.Now().String(), "rabbitmq: connecting to "+host+":"+port)
	connection, err := amqp091.Dial("amqp://" + username + ":" + password + "@" + host + ":" + port + "/")
	if err != nil {
		log.Fatalln("error when connecting rabbitmq: ", err)
	}
	println(time.Now().String(), "rabbitmq: connected to "+host+":"+port)

	println(time.Now().String(), "rabbitmq: opening channel "+host+":"+port)
	channel, err := connection.Channel()
	if err != nil {
		log.Fatalln("error when opening channel:", err)
	}
	println(time.Now().String(), "rabbitmq: opened channel "+host+":"+port)

	return &rabbitmqUtil{
		Connection: connection,
		Channel:    channel,
	}
}

func (util *rabbitmqUtil) GetConnection() *amqp091.Connection {
	return util.Connection
}

func (util *rabbitmqUtil) GetChannel() *amqp091.Channel {
	return util.Channel
}

func (util *rabbitmqUtil) Close(host string, port string) {
	println(time.Now().String(), "rabbitmq: closing to "+host+":"+port)
	util.Connection.Close()
	println(time.Now().String(), "rabbitmq: closed to "+host+":"+port)
}

File: rabbitmq/net9/Utils/RabbitMQUtil.cs
using RabbitMQ.Client;
using RabbitMQ.Client.Events;


namespace net9.Utils;

public class RabbitMQUtil: IAsyncDisposable
{
    private readonly IConnectionFactory _connectionFactory;
    private IConnection _connection;
    private IChannel _channel;

    public RabbitMQUtil()
    {
        _connectionFactory = new ConnectionFactory()
        {
            HostName = "localhost",
            UserName = "user",
            Password = "password",
            Port = 5672
        };
    }

    public async Task InitializeAsync()
    {
        Console.WriteLine($"{DateTime.UtcNow.ToString("yyyy-MM-dd HH:mm:ss")} rabbitmq: connecting to localhost:5672");
        _connection = await _connectionFactory.CreateConnectionAsync();
        Console.WriteLine($"{DateTime.UtcNow.ToString("yyyy-MM-dd HH:mm:ss")} rabbitmq: connected to localhost:5672");

        Console.WriteLine($"{DateTime.UtcNow.ToString("yyyy-MM-dd HH:mm:ss")} rabbitmq: creating channel");
        _channel = await _connection.CreateChannelAsync();
        Console.WriteLine($"{DateTime.UtcNow.ToString("yyyy-MM-dd HH:mm:ss")} rabbitmq: created channel");

        Console.WriteLine($"{DateTime.UtcNow.ToString("yyyy-MM-dd HH:mm:ss")} rabbitmq: creating queue: notification");
        await _channel.QueueDeclareAsync(queue: "notification", durable: false, exclusive: false, autoDelete: false, arguments: null);
        Console.WriteLine($"{DateTime.UtcNow.ToString("yyyy-MM-dd HH:mm:ss")} rabbitmq: creating queue: notification");
    }

    public IChannel GetChannel() => _channel ?? throw new InvalidOperationException("Channel not initialized");
    
    public async ValueTask DisposeAsync()
    {
        Console.WriteLine($"{DateTime.UtcNow.ToString("yyyy-MM-dd HH:mm:ss")} rabbitmq: closing to localhost:5672");
        await _connection.CloseAsync();
        await _connection.DisposeAsync();
        Console.WriteLine($"{DateTime.UtcNow.ToString("yyyy-MM-dd HH:mm:ss")} rabbitmq: closed to localhost:5672");
    }
}